### ISSUE TITLE
Tutorial Request

### ISSUE THREAD

**Issue by CraigWatt** [ethanv2/uwp#1]
Hi Ethan!

I've been following along your Youtube C tutorials, looking forward to future releases.

I've jumped the gun a bit and nearly got caught up in cmake(also autotools), but I've been looking closely at this projects Makefile and config.mk files.

I've noticed you have decided to stick to manually creating these.

I'm also a big fan of the 'components/modules/library' set up here (I am coming from a web dev background)

So seeing things broken down into 

uwp-help
uwp-test
uwp-post etc..

with uwp being a wrapper...

Is great and makes sense in my brain.

I'll be honest, I really wanted to hit the ground running perhaps using a TDD framework such as GTest, but setting up cmake with GTest, Makefiles etc etc...

Has kind of thrown be a bit.  I'm just one of those that really likes to have a solid project structure, testing process, build process (and eventually CI/CD process) plan 'locked in' before soldiering on with the programming...

A little bit of background from me, the reason I have jumped on C is that I developed a rather verbose, monolithic zsh shell script and wanted to switch to C to give myself more control over what exactly I wanted my system 'to do'.  I also really wanted the way I develop to be a bit more **loosely coupled & modular & good seperation of concern**.
So, essentially develop more TDD & bottom up so that then its easier to call these functions from a higher level and in an order that suites the argv being provided by the user.

I'm not saying i'm allergic to refactoring, but i very much want to avoid developing a 'blob' again like I did in zsh.  **Are TDD frameworks overkill for C beginners/intermediates do you think?**  Happy to avoid this arena for now especially if it's better to focus on manual generation of Makefiles first.  **Maybe there is a clean manual way to integrate GTest outside of something like autotools/cmake?**

TLDR:

**1. Is there any chance you could give an overview of your approach to Makefile and config.mk files?
2. Is there a  particular reason why you do not isolate these separate libraries to their own folders?**

Any insight from you is much appreciated!

Craig

**Comment by CraigWatt** [comment]
Just to add, I am currently following along to this tutorial to really try to establish a solid understanding of Makefile:
https://www.youtube.com/watch?v=DtGrdB8wQ_8&t=1274s&ab_channel=GabrielParmer

I've just clocked some of these variables are implicit hence why $(CC) 'just works'.
https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html

**Comment by CraigWatt** [comment]
I've just clocked some of these variables are implicit hence why $(CC) 'just works'.

**Comment by ejv2** [comment]
Hi Craig.

> **1. Is there any chance you could give an overview of your approach to Makefile and config.mk files?**

Yes, but it wouldn't take a video. My approach is to keep it simple. If you find yourself relying on too many GNU extensions etc. - you have made it too complex. This is possibly a symptom of the fact that this is a relatively simple codebase (of the professional codebases I have seen or worked on, the only one which didn't use cmake was written in PHP). My approach is inspired by the style guidelines set out on <https://suckless.org/coding_style>.

> 2. Is there a particular reason why you do not isolate these separate libraries to their own folders?

Not particularly - in fact, I initially did have them in separate directories and then moved them. The reason for this is that I simply had a makefile in each directory, but each just had the same configs and flags and so on. Combined with the fact that the source code for each tool is effectively one file anyways (all shared code is already compiled to separate object files), I thought that directories and sub-makefiles were overkill. So, I decided that I would be achieving the same effect by simply having everything in one directory and just moved it there.

``config.mk`` is designed for use as a build-time configuration file for (among others) package maintainers or distributors who are configuring software for their specific use case.

> I'm just one of those that really likes to have a solid project structure, testing process, build process (and eventually CI/CD process) plan 'locked in' before soldiering on with the programming...

My ordering for C projects is:

1. Create a main file
2. Get it building using a basic makefile
3. Expand and repeat the above as the codebase grows

Because it is quite easy to expand your build system, but it is quite irritating if your build system is massive overkill because your project ended up small.

> **Are TDD frameworks overkill for C beginners/intermediates do you think?**

In all honesty, probably. Unless your project is enormous, modular and sprawling, you probably don't really need to write it based on a bunch of test cases. Just incrementally test your code as you go along and get a minimum product. Then, improve from there. I never sit down and write up a conclusive design for a project unless I anticipate the project being quite large. For examples of projects I planned out, see DSJAS (https://github.com/DSJAS/DSJAS) gahoot (https://github.com/ethanv2/gahoot) and podbit (https://github.com/ethanv2/podbit). In the case of podbit, regardless of having a full design document before starting, the project still turned out to be quite messy. I say that it is best to just get writing and then design components on top of what you build as ground work.

Thanks and good luck with your rewrite.

**Comment by CraigWatt** [comment]
Hi again @ethanv2 

Big thanks for your response above, it helped keep me on the right track!

Wanted to give you an update:

I am happy to say I've finally made my first proper C program available: https://github.com/CraigWatt/vfo

Everything is up and running as far as being appropriate for a v0.1.0 (I wouldn't go as far as 1.0.0 release yet until I'm 100%, but for us 'dev types' anyway, the MVP is 'working' IMO.

Final big piece of the puzzle imo is Delivery/Continuous Delivery so that installation is a super smooth process for any end user (Linux Distro OR Mac OR Windows).

But anyway at least the repo is up now!  I'm going to do a video walkthrough of the basic use of the program, and a video going in depth on the code/development up to this point and how it all works.

I'll gladly link to your C tutorials as inspiration for getting me back into C development!

All the best,

Craig
